File: src/semp_workflow/modules/queue.py
"""Queue module - queue.add, queue.delete, queue.update."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient
from ..semp.helpers import check_name_length, clean_payload, coerce_bool, coerce_int, enc
from .base import BaseModule

logger = logging.getLogger(__name__)

BOOL_FIELDS = {"ingressEnabled", "egressEnabled"}
INT_FIELDS = {"maxMsgSpoolUsage", "maxTtl", "maxRedeliveryCount"}


def _build_queue_payload(args: dict) -> dict:
    """Build a SEMP queue payload with proper type coercion."""
    payload = clean_payload(args)
    for key in BOOL_FIELDS:
        if key in payload:
            payload[key] = coerce_bool(payload[key])
    for key in INT_FIELDS:
        if key in payload:
            payload[key] = coerce_int(payload[key])
    # Derive respectTtlEnabled from maxTtl: 0 disables it, any positive value enables it
    if "maxTtl" in payload:
        payload["respectTtlEnabled"] = payload["maxTtl"] > 0
    if "maxRedeliveryCount" in payload:
        # -1 is a sentinel meaning "disable redelivery and set count to 0"
        if payload["maxRedeliveryCount"] == -1:
            payload["maxRedeliveryCount"] = 0
            payload["redeliveryEnabled"] = False
        else:
            payload["redeliveryEnabled"] = (payload["maxRedeliveryCount"] != 0)
    return payload


class QueueAdd(BaseModule):
    description = "Create a queue on the message VPN. Skipped if the queue already exists."
    params = {
        "queueName":                          {"type": "string",  "required": True,  "description": "Name of the queue to create"},
        "accessType":                         {"type": "string",  "required": False, "description": "Message delivery pattern", "default": "exclusive", "enum": ["exclusive", "non-exclusive"]},
        "owner":                              {"type": "string",  "required": False, "description": "Client username that owns the queue"},
        "permission":                         {"type": "string",  "required": False, "description": "Permission for non-owner clients", "default": "no-access", "enum": ["no-access", "read-only", "consume", "modify-topic", "delete"]},
        "deadMsgQueue":                       {"type": "string",  "required": False, "description": "Name of the dead-message queue for undeliverable messages"},
        "maxMsgSpoolUsage":                   {"type": "integer", "required": False, "description": "Maximum spool usage in MB (0 = unlimited)"},
        "maxTtl":                             {"type": "integer", "required": False, "description": "Maximum time-to-live for messages in seconds. 0 disables TTL enforcement; any positive value enables it automatically."},
        "ingressEnabled":                     {"type": "boolean", "required": False, "description": "Allow clients to send messages to the queue", "default": "true"},
        "egressEnabled":                      {"type": "boolean", "required": False, "description": "Allow clients to consume messages from the queue", "default": "true"},
        "maxRedeliveryCount":                 {"type": "integer", "required": False, "description": "Maximum redelivery attempts before routing to DMQ (0 = unlimited)"},
        "rejectMsgToSenderOnDiscardBehavior": {"type": "string",  "required": False, "description": "Action when a message is discarded", "enum": ["never", "when-queue-enabled", "always"]},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        queue_name = args.get("queueName", "")
        if not queue_name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: queueName")

        if err := check_name_length("queueName", queue_name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"queues/{enc(queue_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking queue: {e}")

        if exists:
            return ActionResult(ResultStatus.SKIPPED, f"Queue '{queue_name}' already exists")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would create queue '{queue_name}'")

        try:
            payload = _build_queue_payload(args)
            client.create("queues", payload)
            return ActionResult(ResultStatus.OK, f"Queue '{queue_name}' created")
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to create queue '{queue_name}': {e}")


class QueueDelete(BaseModule):
    description = "Delete a queue from the message VPN. Skipped if the queue does not exist."
    params = {
        "queueName": {"type": "string", "required": True, "description": "Name of the queue to delete"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        queue_name = args.get("queueName", "")
        if not queue_name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: queueName")

        if err := check_name_length("queueName", queue_name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"queues/{enc(queue_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking queue: {e}")

        if not exists:
            return ActionResult(ResultStatus.SKIPPED, f"Queue '{queue_name}' does not exist")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would delete queue '{queue_name}'")

        try:
            client.delete(path)
            return ActionResult(ResultStatus.OK, f"Queue '{queue_name}' deleted")
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to delete queue '{queue_name}': {e}")


class QueueUpdate(BaseModule):
    description = "Update attributes of an existing queue. Fails if the queue does not exist."
    params = {
        "queueName":                          {"type": "string",  "required": True,  "description": "Name of the queue to update"},
        "accessType":                         {"type": "string",  "required": False, "description": "Message delivery pattern", "enum": ["exclusive", "non-exclusive"]},
        "owner":                              {"type": "string",  "required": False, "description": "Client username that owns the queue"},
        "permission":                         {"type": "string",  "required": False, "description": "Permission for non-owner clients", "enum": ["no-access", "read-only", "consume", "modify-topic", "delete"]},
        "deadMsgQueue":                       {"type": "string",  "required": False, "description": "Name of the dead-message queue for undeliverable messages"},
        "maxMsgSpoolUsage":                   {"type": "integer", "required": False, "description": "Maximum spool usage in MB (0 = unlimited)"},
        "maxTtl":                             {"type": "integer", "required": False, "description": "Maximum time-to-live for messages in seconds. 0 disables TTL enforcement; any positive value enables it automatically."},
        "ingressEnabled":                     {"type": "boolean", "required": False, "description": "Allow clients to send messages to the queue"},
        "egressEnabled":                      {"type": "boolean", "required": False, "description": "Allow clients to consume messages from the queue"},
        "maxRedeliveryCount":                 {"type": "integer", "required": False, "description": "Maximum redelivery attempts before routing to DMQ (0 = unlimited)"},
        "rejectMsgToSenderOnDiscardBehavior": {"type": "string",  "required": False, "description": "Action when a message is discarded", "enum": ["never", "when-queue-enabled", "always"]},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        queue_name = args.get("queueName", "")
        if not queue_name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: queueName")

        if err := check_name_length("queueName", queue_name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"queues/{enc(queue_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking queue: {e}")

        if not exists:
            return ActionResult(ResultStatus.FAILED, f"Queue '{queue_name}' does not exist")

        payload = _build_queue_payload(args)
        payload.pop("queueName", None)

        if not payload:
            return ActionResult(ResultStatus.SKIPPED, "No fields to update")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would update queue '{queue_name}'")

        try:
            client.update(path, payload)
            return ActionResult(ResultStatus.OK, f"Queue '{queue_name}' updated")
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to update queue '{queue_name}': {e}")


MODULES = {
    "queue.add": QueueAdd,
    "queue.delete": QueueDelete,
    "queue.update": QueueUpdate,
}

File: internal/modules/queue.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("queue.add", queueAdd{})
	register("queue.delete", queueDelete{})
	register("queue.update", queueUpdate{})
}

// buildQueuePayload builds a SEMP queue payload with type coercion and derives
// respectTtlEnabled/redeliveryEnabled.
func buildQueuePayload(args map[string]any) (map[string]any, error) {
	payload := semp.CleanPayload(args)
	coerceBoolFields(payload, "ingressEnabled", "egressEnabled")
	if err := coerceIntFields(payload, "maxMsgSpoolUsage", "maxTtl", "maxRedeliveryCount"); err != nil {
		return nil, err
	}
	// Derive respectTtlEnabled from maxTtl: 0 disables it, positive enables it.
	if v, ok := payload["maxTtl"]; ok {
		payload["respectTtlEnabled"] = v.(int) > 0
	}
	if v, ok := payload["maxRedeliveryCount"]; ok {
		// -1 is a sentinel meaning "disable redelivery and set count to 0".
		if v.(int) == -1 {
			payload["maxRedeliveryCount"] = 0
			payload["redeliveryEnabled"] = false
		} else {
			payload["redeliveryEnabled"] = v.(int) != 0
		}
	}
	return payload, nil
}

type queueAdd struct{}

func (queueAdd) Description() string {
	return "Create a queue on the message VPN. Skipped if the queue already exists."
}

func (queueAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "queueName", Type: "string", Required: true, Description: "Name of the queue to create"},
		{Name: "accessType", Type: "string", Description: "Message delivery pattern", Default: "exclusive", Enum: []string{"exclusive", "non-exclusive"}},
		{Name: "owner", Type: "string", Description: "Client username that owns the queue"},
		{Name: "permission", Type: "string", Description: "Permission for non-owner clients", Default: "no-access", Enum: []string{"no-access", "read-only", "consume", "modify-topic", "delete"}},
		{Name: "deadMsgQueue", Type: "string", Description: "Name of the dead-message queue for undeliverable messages"},
		{Name: "maxMsgSpoolUsage", Type: "integer", Description: "Maximum spool usage in MB (0 = unlimited)"},
		{Name: "maxTtl", Type: "integer", Description: "Maximum time-to-live for messages in seconds. 0 disables TTL enforcement; any positive value enables it automatically."},
		{Name: "ingressEnabled", Type: "boolean", Description: "Allow clients to send messages to the queue", Default: "true"},
		{Name: "egressEnabled", Type: "boolean", Description: "Allow clients to consume messages from the queue", Default: "true"},
		{Name: "maxRedeliveryCount", Type: "integer", Description: "Maximum redelivery attempts before routing to DMQ (0 = unlimited)"},
		{Name: "rejectMsgToSenderOnDiscardBehavior", Type: "string", Description: "Action when a message is discarded", Enum: []string{"never", "when-queue-enabled", "always"}},
	}
}

func (queueAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	queueName := argStr(args, "queueName", "")
	if queueName == "" {
		return failed("Missing required arg: queueName")
	}
	if err := semp.CheckNameLength("queueName", queueName); err != "" {
		return failed(err)
	}

	path := "queues/" + semp.Enc(queueName)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking queue: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("Queue '%s' already exists", queueName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would create queue '%s'", queueName))
	}

	payload, err := buildQueuePayload(args)
	if err != nil {
		return failed(err.Error())
	}
	if _, err := c.Create("queues", payload); err != nil {
		return failed(fmt.Sprintf("Failed to create queue '%s': %s", queueName, err))
	}
	return ok(fmt.Sprintf("Queue '%s' created", queueName))
}

type queueDelete struct{}

func (queueDelete) Description() string {
	return "Delete a queue from the message VPN. Skipped if the queue does not exist."
}

func (queueDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "queueName", Type: "string", Required: true, Description: "Name of the queue to delete"},
	}
}

func (queueDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	queueName := argStr(args, "queueName", "")
	if queueName == "" {
		return failed("Missing required arg: queueName")
	}
	if err := semp.CheckNameLength("queueName", queueName); err != "" {
		return failed(err)
	}

	path := "queues/" + semp.Enc(queueName)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking queue: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("Queue '%s' does not exist", queueName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would delete queue '%s'", queueName))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to delete queue '%s': %s", queueName, err))
	}
	return ok(fmt.Sprintf("Queue '%s' deleted", queueName))
}

type queueUpdate struct{}

func (queueUpdate) Description() string {
	return "Update attributes of an existing queue. Fails if the queue does not exist."
}

func (queueUpdate) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "queueName", Type: "string", Required: true, Description: "Name of the queue to update"},
		{Name: "accessType", Type: "string", Description: "Message delivery pattern", Enum: []string{"exclusive", "non-exclusive"}},
		{Name: "owner", Type: "string", Description: "Client username that owns the queue"},
		{Name: "permission", Type: "string", Description: "Permission for non-owner clients", Enum: []string{"no-access", "read-only", "consume", "modify-topic", "delete"}},
		{Name: "deadMsgQueue", Type: "string", Description: "Name of the dead-message queue for undeliverable messages"},
		{Name: "maxMsgSpoolUsage", Type: "integer", Description: "Maximum spool usage in MB (0 = unlimited)"},
		{Name: "maxTtl", Type: "integer", Description: "Maximum time-to-live for messages in seconds. 0 disables TTL enforcement; any positive value enables it automatically."},
		{Name: "ingressEnabled", Type: "boolean", Description: "Allow clients to send messages to the queue"},
		{Name: "egressEnabled", Type: "boolean", Description: "Allow clients to consume messages from the queue"},
		{Name: "maxRedeliveryCount", Type: "integer", Description: "Maximum redelivery attempts before routing to DMQ (0 = unlimited)"},
		{Name: "rejectMsgToSenderOnDiscardBehavior", Type: "string", Description: "Action when a message is discarded", Enum: []string{"never", "when-queue-enabled", "always"}},
	}
}

func (queueUpdate) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	queueName := argStr(args, "queueName", "")
	if queueName == "" {
		return failed("Missing required arg: queueName")
	}
	if err := semp.CheckNameLength("queueName", queueName); err != "" {
		return failed(err)
	}

	path := "queues/" + semp.Enc(queueName)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking queue: %s", err))
	}
	if !exists {
		return failed(fmt.Sprintf("Queue '%s' does not exist", queueName))
	}

	payload, err := buildQueuePayload(args)
	if err != nil {
		return failed(err.Error())
	}
	delete(payload, "queueName")

	if len(payload) == 0 {
		return skipped("No fields to update")
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would update queue '%s'", queueName))
	}

	if _, err := c.Update(path, payload); err != nil {
		return failed(fmt.Sprintf("Failed to update queue '%s': %s", queueName, err))
	}
	return ok(fmt.Sprintf("Queue '%s' updated", queueName))
}
